### PR TITLE
✨ Add registry provider client and integration tests

### PR DESCRIPTION
**Pull request by annawinkler**
## Description

This PR was broken out of this [PR](https://github.com/hashicorp/go-tfe/pull/313).

This PR adds support for the private registry provider API.

## Testing plan

To test these changes locally, create a local go project, and in the `go.mod` include a line similar to the following:

```
replace github.com/hashicorp/go-tfe => /Users/YOURUSER/go/src/github.com/hashicorp/go-tfe
```

Then, for a registry provider, you can test:
* Create
* Delete
* Read
* List

## External links

- [Registry Providers API](https://www.terraform.io/cloud-docs/api-docs/private-registry/providers)
- [Private Registry Provider Requirements](https://github.com/hashicorp/go-tfe/pull/399)

## Output from tests
Including output from tests may require access to a TFE instance. Ignore this section if you have no environment to test against.

```
$ TFE_ADDRESS="https://example" TFE_TOKEN="example" TF_ACC="1" go test ./... -v -tags=integration -run TestFunctionsAffectedByChange

...
```

### CODE REVIEW

File: registry_provider.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+	*Pagination
+	Items []*RegistryProvider
+}
+
+func (o RegistryProviderListOptions) valid() error {
+	return nil
+}
+
+func (r *registryProviders) List(ctx context.Context, organization string, options *RegistryProviderListOptions) (*RegistryProviderList, error) {
+
```
Comment (sebasslash) [inline registry_provider.go:84]: unnecessary whitespace ✏️

File: registry_provider.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+// RegistryProviderCreateOptions is used when creating a registry provider
+type RegistryProviderCreateOptions struct {
+	// Type is a public field utilized by JSON:API to
+	// set the resource type via the field tag.
+	// It is not a user-defined value and does not need to be set.
+	// https://jsonapi.org/format/#crud-creating
+	Type string `jsonapi:"primary,registry-providers"`
+
+	Namespace    *string       `jsonapi:"attr,namespace"`
```
Comment (sebasslash) [inline registry_provider.go:117]: These fields should not be pointers unless the field is a struct type or the API accepts a `null` value (and perhaps that may be so since `omitempty` is omitted?)

File: registry_provider.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+func (r *registryProviders) Create(ctx context.Context, organization string, options RegistryProviderCreateOptions) (*RegistryProvider, error) {
+	if !validStringID(&organization) {
+		return nil, ErrInvalidOrg
+	}
+	if err := options.valid(); err != nil {
+		return nil, err
+	}
+	// Private providers must match their namespace and organization name
+	// This is enforced by the API as well
+	if *options.RegistryName == PrivateRegistry && organization != *options.Namespace {
```
Comment (sebasslash) [inline registry_provider.go:150]: this check should live inside the `RegistryProviderCreateOptions.valid()` function

File: registry_provider.go [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+	prv := &RegistryProvider{}
+	err = r.client.do(ctx, req, prv)
+	if err != nil {
+		return nil, err
+	}
+
+	return prv, nil
+}
+
+type RegistryProviderReadOptions struct {
```
Comment (sebasslash) [inline registry_provider.go:171]: Empty struct? 👻

File: registry_provider.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+func (r *registryProviders) Read(ctx context.Context, organization string, registryName RegistryName, namespace string, name string, options *RegistryProviderReadOptions) (*RegistryProvider, error) {
```
Comment (sebasslash) [inline registry_provider.go:174]: Whenever I see a long method signature like that I wonder if we can use a struct to keep the signature concise? An example:
https://github.com/hashicorp/go-tfe/blob/11a4eb9078cef3c9caf389325ddc6327e749ea94/registry_module.go#L75-L82

File: registry_provider.go [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+	if !validStringID(&organization) {
```
Comment (sebasslash) [inline registry_provider.go:175]: All these `validStringID()` checks should probably be refactored into a `valid()` function and perhaps should live in the `ReadOptionsStruct` instead of as method params (back to my previous point). What do you think?

File: registry_provider_integration_test.go [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+		defer orgTestCleanup()
+
+		createN := 10
+		providers := make([]*RegistryProvider, 0)
+		// these providers will be destroyed when the org is cleaned up
+		for i := 0; i < createN; i++ {
+			providerTest, _ := createPublicRegistryProvider(t, client, orgTest)
+			providers = append(providers, providerTest)
+		}
+		for i := 0; i < createN; i++ {
```
Comment (sebasslash) [inline registry_provider_integration_test.go:27]: Duplicate for loop?

File: registry_provider_integration_test.go [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+				ListOptions: ListOptions{
+					PageNumber: 0,
+					PageSize:   providerN,
+				},
+			})
+			require.NoError(t, err)
+			assert.NotEmpty(t, returnedProviders.Items)
+			assert.Equal(t, providerN, returnedProviders.TotalCount)
+			assert.Equal(t, 1, returnedProviders.TotalPages)
+			for _, rp := range returnedProviders.Items {
```
Comment (sebasslash) [inline registry_provider_integration_test.go:45]: Alternatively, we can condense this logic to:

```go
for _, p := range providers {
  assert.Contains(t, returnedProviders.Items, p)
}
```

Careful with `assert.Contains()` as it performs deep equality. (I've been burned before with `Included` relationships before 👉 🔥 )

Comment (sebasslash) [reply registry_provider_integration_test.go:45]: Furthermore, this sort of logic (checking that the desired provider(s) exists) seems better reserved for the `Create` test; for this test we should only care about that endpoint returns the correct # of elements, query params work as intended and that the fields are deserialized properly (although we can reserve checking the deserialization of fields in the `Read` test).

File: registry_provider_integration_test.go [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+					if rp.ID == p.ID {
+						foundProvider = true
+						break
+					}
+				}
+				assert.True(t, foundProvider, "Expected to find provider %s but did not:\nexpected:\n%v\nreturned\n%v", rp.ID, providers, returnedProviders)
+			}
+		})
+
+		t.Run("returns pages", func(t *testing.T) {
```
Comment (sebasslash) [inline registry_provider_integration_test.go:57]: For this kind of test, the usual convention is to set `PageSize` and `PageNumber` to some incredibly large number, which I think is the easiest way to verify pagination is working as intended. Here's an example:

https://github.com/hashicorp/go-tfe/blob/11a4eb9078cef3c9caf389325ddc6327e749ea94/workspace_integration_test.go#L44-L58

File: registry_provider_integration_test.go [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+				assert.Equal(t, publicName, rp.RegistryName)
+				assert.True(t, foundProvider, "Expected to find provider %s but did not:\nexpected:\n%v\nreturned\n%v", rp.ID, providers, returnedProviders)
+			}
+		})
+
+		t.Run("searches", func(t *testing.T) {
+			expectedProvider := providers[0]
+			returnedProviders, err := client.RegistryProviders.List(ctx, orgTest.Name, &RegistryProviderListOptions{
+				Search: &expectedProvider.Name,
+				ListOptions: ListOptions{
```
Comment (sebasslash) [inline registry_provider_integration_test.go:119]: Can we omit the ListOptions field here?

**Review by sebasslash — CHANGES_REQUESTED** [review]
Great work @annawinkler and @joekarl ... I have not been able to review all of it as it is quite a hefty PR, but hopefully a lot the comments translate to the rest of the unreviewed code. 

Files I've reviewed:
helper_test.go
registry_provider.go
And the `List` test in the registry_provider_integration_test.go

Definitely would like @brandonc and @uturunku1 to give it a pass.

File: registry_provider.go (hunk 5 above)
Comment (annawinkler) [reply registry_provider.go:169]: This is refactored in the next PR. Wondering if it's simpler to go with the large PR that has everything?

**Comment by annawinkler** [comment]
Closing this PR in favor of the one large [one](https://github.com/hashicorp/go-tfe/pull/313) since refactoring has occurred there. I'll address the review comments in PR 313.

File: registry_provider.go (hunk 2 above)
Comment (annawinkler) [reply registry_provider.go:117]: In PR https://github.com/hashicorp/go-tfe/pull/313 they are not pointers.

File: registry_provider.go (hunk 6 above)
Comment (annawinkler) [reply registry_provider.go:170]: This has been refactored.

File: registry_provider_integration_test.go (hunk 7 above)
Comment (annawinkler) [reply registry_provider_integration_test.go:27]: It looks like it but we create private and public registry providers in the test.
